### ISSUE TITLE
How can I call getPort() and just have it return an open port?

### ISSUE THREAD

**Issue by axfelix** [indexzero/node-portfinder#76]
Hi,

I'm afraid this is an extremely stupid question, but I'm coming at this with a Pythonic brain and I can't figure out from your documentation how I'm supposed to return an openPort from a call to getPort without having to rearrange all of my code inside of its scope.

**Comment by eriktrom** [comment]
```
portfinder.getPort(function(err, port) {
  if (err) {   return console.log(err); } 
  console.log(port);
});
```

There is also a promisified version

```js
return getPortPromise((port) => {
  console.log(port);
})
.catch((err) => {
  console.log(err);
});
```

also worth noting the optional options arg  `{ host: string, port: <string|number> }` can be provided as well, as first arg to either version. The host will be added to the default hosts we check (::1, localhost, 0.0.0.0), and the port option will start iterating through ports starting with that port..

**Comment by axfelix** [comment]
Sorry, I should've clarified -- in the non-promisified version, what's the easiest way of returning `port` outside of that scope? That's the part that I'm just not getting.

**Comment by eriktrom** [comment]
sorry for the delay!

tl;dr - with async callbacks in node - you don't :)

you might think u **could** do something like the following (but dont!)... i'll throw a python async like way at the bottom... 

but first, callback hell style with common gotcha when trying to *get around* it:

```js
let port;

portfinder.getPort(function(err, port) {
  if (err) {   return console.log(err); } 
  port = port;
});

console.log(port) // => undefined
```

the problem with this is the same problem with this code:

```js
setTimeout(() => {
  console.log('will log 2nd');
}, 0);

console.log('will log first');
```

e.g., there is no way to deterministically know that the port outside of the callback scope will be set, or when it will be set (in reality opening a socket, testing the port, closing the socket is pretty slow). You have to write all code that uses the port value, returned asynchronously via the callback, within the scope of that callback... which is why they call it callback hell b/c when u have many async callbacks your code indents rightword as you call function after function that all return values via another async callback... i know right... 

Silver Lining: (python inspired too)
 
if u can (meaning team/current code makes it possible), do it like python's async fn's

```js
async doSomething() {
  let port = await getPortPromise(); // pass options if you want. like before, must be a promise that resolves w/ a value, code will not continue until async promise is resolved
  // do something with the port, no callback hell or promise resolution syntax needed, feels sync, fun!
}
```

and b/c i feel ur pain.. one more thing - you can take any callback style async node code and use the util.promisify(functionName) and get a promisified version that allows u to use async await - check the nodejs docs util module(node 8 or 10, dont recall) - and if u need it in an earlier versions of node, checkout the library called bluebird.js which has similar promisify utility helper fn for earlier versions of node.

good luck, it'll grow on u, not as complete a lang as python, but es6 takes a lot from python, it's changing the language for the better

cheerio

**Comment by axfelix** [comment]
Thanks so much!
